lib/scholar/decomposition/pca.ex

```diff
@@ -0,0 +1,215 @@
+
+    * `:singular_values` - The singular values corresponding to each of the selected components.
+      The singular values are equal to the 2-norms of the `:num_components` variables in the lower-dimensional space.
+
+    * `:mean` - Per-feature empirical mean, estimated from the training set.
+
+    * `:num_components` - It equals the parameter `:num_components`, or the lesser
+      value of `:num_features` and `:num_samples` if the parameter `:num_components` is `nil`.
+
+    * `:num_features` - Number of features in the training data.
+
+    * `:num_samples` - Number of samples in the training data.
+  """
+  deftransform fit(x, opts \\ []) do
+    fit_n(x, NimbleOptions.validate!(opts, @fit_opts_schema))
+  end
+
+  # TODO Add support for :num_components as a float when dynamic shapes will be implemented
+  defnp fit_n(x, opts \\ []) do
+    if Nx.rank(x) != 2 do
+      raise ArgumentError, "expected x to have rank equal to: 2, got: #{inspect(Nx.rank(x))}"
+    end
+
+    {num_samples, num_features} = Nx.shape(x)
+    num_components = opts[:num_components]
+
+    mean = Nx.mean(x, axes: [0])
+    x = x - mean
+    {decomposer, singular_values, components} = Nx.LinAlg.svd(x)
+
+    num_components =
+      calculate_num_components(
+        num_components,
+        num_features,
+        num_samples
+      )
+
+    trim_dim = min_shape(num_samples, num_features)
+
+    {_, components} = flip_svd(decomposer[[0..-1//1, 0..(trim_dim - 1)]], components)
+
+    explained_variance = singular_values * singular_values / (num_samples - 1)
+    explained_variance_ratio = explained_variance / Nx.sum(explained_variance)
+
+    %__MODULE__{
+      components: components,
+      explained_variance: explained_variance,
+      explained_variance_ratio: explained_variance_ratio,
+      singular_values: singular_values,
+      mean: mean,
+      num_components: num_components,
+      num_features: num_features,
+      num_samples: num_samples
+    }
+  end
+
+  @doc """
+  For a fitted model performs a decomposition.
+
+  ## Options
+
+  #{NimbleOptions.docs(@transform_opts_schema)}
+
+  ## Returns
+
+  The function returns a decomposed data.
+  """
+  deftransform transform(model, x, opts \\ []) do
+    transform_n(model, x, NimbleOptions.validate!(opts, @transform_opts_schema))
+  end
+
+  defnp transform_n(
+          %__MODULE__{
+            components: components,
+            explained_variance: explained_variance,
+            mean: mean,
+            num_components: num_components
+          } = _model,
+          x,
+          opts \\ []
+        ) do
+    whiten? = opts[:whiten]
+
+    x = x - mean
+
+    x_transformed =
+      Nx.dot(
+        x,
+        Nx.transpose(components)
+      )
+
+    x_transformed =
+      if not whiten? do
+        x_transformed
+      else
+        x_transformed / Nx.sqrt(explained_variance)
+      end
+
+    x_transformed[[0..-1//1, 0..(num_components - 1)]]
```

Comment:
We should slice the model in fit, such that we only have `:num_components` components (and the other fields accordingly). This way we don't need to unnecessarily compute and slice on transform.